### ISSUE TITLE
Location can't be retrieved

### ISSUE THREAD

**Issue by AlexandrSushkov** [appoly/ARCore-Location#9]
Hi guys, couple days ago I have found your library and tested your sample, everything worked well but now I am getting this error "Location can't be retrieved". What can be a problem, how can I fix it?

**Comment by johnwedgbury** [comment]
Hi alexandr,This sometimes happens when your phone hasn't yet received any location updates since booting. Opening Google maps before rubbing your project usually fixes this.It's an issue I'm aware of, but let me know if you have any other issues.On 4 May 2018 18:01, "alexandr.sushkov" <notifications@github.com> wrote:Hi guys, couple days ago I have found your library and tested your sample, everything worked well but now I am getting this error "Location can't be retrieved". What can be a problem, how can I fix it?
java.lang.NullPointerException: Attempt to invoke virtual method 'double android.location.Location.getLatitude()' on a null object reference
at uk.co.appoly.arcorelocation.LocationScene.refreshAnchorsIfRequired(LocationScene.java:162)
at uk.co.appoly.arcorelocation.LocationScene.draw(LocationScene.java:78)
at com.google.ar.core.examples.java.helloar.HelloArActivity.onDrawFrame(HelloArActivity.java:351)
at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1571)
at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1270)
java.lang.NullPointerException: Attempt to invoke virtual method 'com.google.ar.core.Pose com.google.ar.core.Anchor.getPose()' on a null object reference
at uk.co.appoly.arcorelocation.LocationScene.drawMarkers(LocationScene.java:102)
at uk.co.appoly.arcorelocation.LocationScene.draw(LocationScene.java:81)
at com.google.ar.core.examples.java.helloar.HelloArActivity.onDrawFrame(HelloArActivity.java:351)
at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1571)
at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1270)
05-04 18:50:15.113 28122-28168/com.google.ar.core.examples.java.helloar W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'com.google.ar.core.Pose com.google.ar.core.Anchor.getPose()' on a null object reference
at uk.co.appoly.arcorelocation.LocationScene.drawMarkers(LocationScene.java:102)
at uk.co.appoly.arcorelocation.LocationScene.draw(LocationScene.java:81)
at com.google.ar.core.examples.java.helloar.HelloArActivity.onDrawFrame(HelloArActivity.java:351)
at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1571)
at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1270)
05-04 18:50:15.114 28122-28168/com.google.ar.core.examples.java.helloar W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'com.google.ar.core.Pose com.google.ar.core.Anchor.getPose()' on a null object reference
at uk.co.appoly.arcorelocation.LocationScene.drawMarkers(LocationScene.java:102)
at uk.co.appoly.arcorelocation.LocationScene.draw(LocationScene.java:81)
at com.google.ar.core.examples.java.helloar.HelloArActivity.onDrawFrame(HelloArActivity.java:351)
at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1571)
at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1270)

**Comment by repentant** [comment]
Same problem here. I opened Google Maps and waited about 5 minutes but nothing happened. I'm still having "location can't be retrieved" toast. Do you consider any workaround in a near future?

**Comment by rahul-raina3** [comment]
So this library was working fine for the past few days until today I ran into a similar issue. I think part of the issue is that in DeviceLocation the locationManager is trying to use locationManager.getLastLocation("gps") and sometimes the gps isn't super accurate or quick to respond. 

I made a work around in my project by doing this:
locationManager = (LocationManager) getBaseContext().getSystemService(Context.LOCATION_SERVICE);
                if (locationManager != null) {
                    locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 500, 1, locationListener);
                    locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 500, 1, locationListener);
                }

basically requesting location updates form both NETWORK_PROVIDER and GPS_PROVIDER. Then in the locationListener( { onLocationChanged(LocationResult locationResult) } callback, I instantiate locationScene and all of my AR assets.

This basically forces my app to wait until it knows its location before creating a LocationScene instance so that locationScene can know that there is a "getLastLocation" before attempting to instantiate.

It'd be really awesome if some functionality like this could be built in, but for now this is what's working for me.

**Comment by rahul-raina3** [comment]
I'm realizing today that my proposed solution doesn't necessarily help because the DeviceLocation class tries to do locationManager.getLastLocation(GPS_PROVIDER), so getting your location based on NETWORK_PROVIDER by doing locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,...), isn't really any help because you need to get location based off of GPS_PROVIDER. This is problematic because GPS_PROVIDER can be very slow, especially inside buildings. 

If we could get an update to the library accounting for NETWORK_PROVIDER in DeviceLocation, I think it'd be a huge help in terms of getting location faster.

**Comment by johnwedgbury** [comment]
This has been addressed in version 1.0.4 which will attempt to use NETWORK_PROVIDER before GPS. Let me know if this hasn't solved your issues.
